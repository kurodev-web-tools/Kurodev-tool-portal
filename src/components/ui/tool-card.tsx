import React from "react";
import Link from "next/link";
import { Card, CardContent, CardDescription, CardFooter, CardHeader, CardTitle } from "./card";
import { Badge } from "./badge";
import { Button } from "./button";
import { ArrowRight, Star, Users } from "lucide-react";

type ToolStatus = "released" | "beta" | "development";

interface ToolCardProps {
  id: string;
  title: string;
  description: string;
  status: ToolStatus;
  feedbackMessage?: string;
  hoverable?: boolean;
  href?: string;
  iconName?: string;
  color?: string;
  stats?: string;
  onClick?: () => void;
}

// アイコンマッピング
const iconMap = {
  "sparkles": "✨",
  "trending-up": "📈",
  "users": "👥",
  "calendar": "📅",
  "image": "🖼️",
  "brain": "🧠",
  "bar-chart": "📊",
};

export const ToolCard = React.memo(function ToolCard({
  id,
  title,
  description,
  status,
  feedbackMessage,
  hoverable = false,
  href,
  iconName,
  color = "from-gray-500 to-gray-600",
  stats,
  onClick,
}: ToolCardProps) {
  const statusMap = {
    released: { text: "公開中", variant: "default", color: "bg-green-200 text-green-900 border-green-300 dark:bg-green-800 dark:text-green-100 dark:border-green-600" },
    beta: { text: "ベータ版", variant: "warning", color: "bg-yellow-200 text-yellow-900 border-yellow-300 dark:bg-yellow-800 dark:text-yellow-100 dark:border-yellow-600" },
    development: { text: "開発中", variant: "destructive", color: "bg-red-200 text-red-900 border-red-300 dark:bg-red-800 dark:text-red-100 dark:border-red-600" },
  };

  const currentStatus = statusMap[status];
  const iconEmoji = iconName ? iconMap[iconName as keyof typeof iconMap] || "🔧" : "🔧";

  const handleCardClick = () => {
    if (onClick) {
      onClick();
    }
  };

  const CardComponent = (
    <Card 
      className="h-full flex flex-col group card-interactive hover:shadow-blue-500/20 border-gray-800 shadow-sm bg-gray-900/30 hover:bg-gray-900/40 cursor-pointer"
      onClick={handleCardClick}
      role="button"
      tabIndex={0}
      aria-label={`${title} - ${description}`}
      aria-describedby={`${id}-description`}
      onKeyDown={(e) => {
        if (e.key === 'Enter' || e.key === ' ') {
          e.preventDefault();
          handleCardClick();
        }
      }}
    >
      {/* デスクトップ用レイアウト */}
      <CardHeader className="pb-4 flex-shrink-0 hidden md:block">
        <div className="flex items-start justify-between mb-3">
          <div className={`w-12 h-12 rounded-lg bg-gradient-to-r ${color} flex items-center justify-center shadow-lg text-2xl hover-scale-sm`}>
            {iconEmoji}
          </div>
          <Badge className={`text-xs font-medium ${currentStatus.color} status-${status}`}>
            {currentStatus.text}
          </Badge>
        </div>
        <CardTitle className="text-lg md:text-xl font-bold group-hover:text-blue-600 transition-smooth leading-tight tracking-wide">
          {title}
        </CardTitle>
        <CardDescription className="text-xs md:text-sm text-gray-600 dark:text-gray-400 leading-relaxed">
          {description}
        </CardDescription>
      </CardHeader>

      {/* モバイル用レイアウト（2列表示） */}
      <CardHeader className="pb-2 flex-shrink-0 md:hidden">
        <div className="flex items-center justify-between mb-2">
          <div className="flex items-center space-x-3">
            <div className={`w-10 h-10 rounded-lg bg-gradient-to-r ${color} flex items-center justify-center shadow-lg text-xl hover-scale-sm`}>
              {iconEmoji}
            </div>
            <div>
              <CardTitle className="text-sm md:text-base font-bold group-hover:text-blue-600 transition-smooth leading-tight tracking-wide">
                {title}
              </CardTitle>
            </div>
          </div>
          <Badge className={`text-xs font-medium ${currentStatus.color} status-${status}`}>
            {currentStatus.text}
          </Badge>
        </div>
        <CardDescription 
          id={`${id}-description`}
          className="text-xs md:text-sm text-gray-500 dark:text-gray-300 leading-relaxed"
        >
          {description}
        </CardDescription>
      </CardHeader>
      
      {/* デスクトップ用コンテンツ */}
      <CardContent className="flex-grow pb-4 hidden md:block">
        {status === "beta" && feedbackMessage && (
          <div className="bg-yellow-50 dark:bg-yellow-900/20 border border-yellow-200 dark:border-yellow-800 rounded-lg p-3 mb-4 shadow-sm">
            <div className="flex items-center gap-2 mb-1">
              <Star className="h-4 w-4 text-yellow-600" />
              <span className="text-xs md:text-sm font-medium text-yellow-800 dark:text-yellow-200">フィードバック募集中</span>
            </div>
            <p className="text-xs md:text-sm text-yellow-800 dark:text-yellow-200 leading-relaxed">{feedbackMessage}</p>
          </div>
        )}
        
        {stats && (
          <div className="flex items-center gap-2 text-xs md:text-sm text-gray-400 dark:text-gray-300">
            <Users className="h-4 w-4" />
            <span>{stats}</span>
          </div>
        )}
      </CardContent>

      {/* モバイル用コンテンツ（簡略化） */}
      <CardContent className="flex-grow pb-2 md:hidden">
        {status === "beta" && feedbackMessage && (
          <div className="bg-yellow-50 dark:bg-yellow-900/20 border border-yellow-200 dark:border-yellow-800 rounded-lg p-2 mb-2 shadow-sm">
            <div className="flex items-center gap-1">
              <Star className="h-3 w-3 text-yellow-600" />
              <span className="text-xs md:text-sm font-medium text-yellow-800 dark:text-yellow-200">フィードバック募集中</span>
            </div>
          </div>
        )}
      </CardContent>
      
      <CardFooter className="pt-0 flex-shrink-0">
        {href && (
          <>
            {/* デスクトップ用ボタン */}
            <Button 
              variant="ghost" 
              className="w-full hidden md:flex group-hover:bg-blue-50 group-hover:text-blue-600 dark:group-hover:bg-blue-900/20 dark:group-hover:text-blue-400 transition-colors"
              asChild
            >
              <Link href={href} className="flex items-center justify-center gap-2">
                ツールを使用する
                <ArrowRight className="h-4 w-4 group-hover:translate-x-1 transition-smooth" />
              </Link>
            </Button>

            {/* モバイル用ボタン（コンパクト） */}
            <Button 
              variant="ghost" 
              className="w-full md:hidden group-hover:bg-blue-50 group-hover:text-blue-600 dark:group-hover:bg-blue-900/20 dark:group-hover:text-blue-400 transition-colors text-xs md:text-sm py-2 tracking-normal"
              asChild
            >
              <Link href={href} className="flex items-center justify-center gap-1">
                使用する
                <ArrowRight className="h-3 w-3 group-hover:translate-x-1 transition-smooth" />
              </Link>
            </Button>
          </>
        )}
      </CardFooter>
    </Card>
  );

  // hrefがある場合は、外側のLinkでラップせず、Card内のボタンでリンクを処理
  return CardComponent;
});