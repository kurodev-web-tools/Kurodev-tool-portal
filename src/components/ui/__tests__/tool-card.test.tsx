import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ToolCard } from '../tool-card';

// Next.jsのLinkコンポーネントをモック
jest.mock('next/link', () => {
  return ({ children, href }: { children: React.ReactNode; href: string }) => {
    return <a href={href}>{children}</a>;
  };
});

describe('ToolCard', () => {
  const defaultProps = {
    id: 'tool-1',
    title: 'テストツール',
    description: 'これはテスト用のツールです',
    status: 'beta' as const,
    href: '/tools/test-tool',
    iconName: 'calendar',
    color: 'from-blue-500 to-cyan-500',
    stats: '利用可能'
  };

  describe('基本的なレンダリング', () => {
    it('ツールのタイトルが表示される', () => {
      render(<ToolCard {...defaultProps} />);
      
      // デスクトップとモバイルの両方に表示されるので、getAllByTextを使用
      const titles = screen.getAllByText('テストツール');
      expect(titles.length).toBeGreaterThan(0);
    });

    it('ツールの説明が表示される', () => {
      render(<ToolCard {...defaultProps} />);
      
      // デスクトップとモバイルの両方に表示されるので、getAllByTextを使用
      const descriptions = screen.getAllByText('これはテスト用のツールです');
      expect(descriptions.length).toBeGreaterThan(0);
    });

    it('ステータスバッジが表示される', () => {
      render(<ToolCard {...defaultProps} />);
      
      // デスクトップとモバイルの両方に表示されるので、getAllByTextを使用
      const badges = screen.getAllByText('ベータ版');
      expect(badges.length).toBeGreaterThan(0);
    });

    it('統計情報が表示される', () => {
      render(<ToolCard {...defaultProps} />);
      
      expect(screen.getByText('利用可能')).toBeInTheDocument();
    });
  });

  describe('ステータス別の表示', () => {
    it('released状態で正しいバッジが表示される', () => {
      render(<ToolCard {...defaultProps} status="released" />);
      
      const badges = screen.getAllByText('公開中');
      expect(badges.length).toBeGreaterThan(0);
    });

    it('beta状態でフィードバックメッセージが表示される', () => {
      render(
        <ToolCard 
          {...defaultProps} 
          status="beta" 
          feedbackMessage="フィードバックお待ちしております"
        />
      );
      
      // デスクトップとモバイルの両方に表示される
      const feedbackMessages = screen.getAllByText('フィードバック募集中');
      expect(feedbackMessages.length).toBeGreaterThan(0);
    });

    it('development状態で開発中バッジが表示される', () => {
      render(<ToolCard {...defaultProps} status="development" />);
      
      const badges = screen.getAllByText('開発中');
      expect(badges.length).toBeGreaterThan(0);
    });
  });

  describe('インタラクション', () => {
    it('カードをクリックするとonClickが呼ばれる', () => {
      const handleClick = jest.fn();
      render(<ToolCard {...defaultProps} onClick={handleClick} />);
      
      const card = screen.getByRole('button');
      fireEvent.click(card);
      
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('EnterキーでもonClickが呼ばれる', () => {
      const handleClick = jest.fn();
      render(<ToolCard {...defaultProps} onClick={handleClick} />);
      
      const card = screen.getByRole('button');
      fireEvent.keyDown(card, { key: 'Enter' });
      
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('SpaceキーでもonClickが呼ばれる', () => {
      const handleClick = jest.fn();
      render(<ToolCard {...defaultProps} onClick={handleClick} />);
      
      const card = screen.getByRole('button');
      fireEvent.keyDown(card, { key: ' ' });
      
      expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it('その他のキーではonClickが呼ばれない', () => {
      const handleClick = jest.fn();
      render(<ToolCard {...defaultProps} onClick={handleClick} />);
      
      const card = screen.getByRole('button');
      fireEvent.keyDown(card, { key: 'a' });
      
      expect(handleClick).not.toHaveBeenCalled();
    });
  });

  describe('アクセシビリティ', () => {
    it('適切なaria-labelが設定される', () => {
      render(<ToolCard {...defaultProps} />);
      
      const card = screen.getByRole('button');
      expect(card).toHaveAttribute('aria-label', 'テストツール - これはテスト用のツールです');
    });

    it('説明のIDが正しく設定される', () => {
      render(<ToolCard {...defaultProps} />);
      
      const card = screen.getByRole('button');
      expect(card).toHaveAttribute('aria-describedby', 'tool-1-description');
    });

    it('tabIndexが設定される', () => {
      render(<ToolCard {...defaultProps} />);
      
      const card = screen.getByRole('button');
      expect(card).toHaveAttribute('tabIndex', '0');
    });
  });

  describe('リンク', () => {
    it('hrefが提供された場合、リンクボタンが表示される', () => {
      render(<ToolCard {...defaultProps} />);
      
      // デスクトップ用のボタン
      const link = screen.getByText('ツールを使用する').closest('a');
      expect(link).toHaveAttribute('href', '/tools/test-tool');
    });

    it('hrefがない場合、リンクボタンは表示されない', () => {
      const { container } = render(<ToolCard {...defaultProps} href={undefined} />);
      
      expect(screen.queryByText('ツールを使用する')).not.toBeInTheDocument();
    });
  });

  describe('スナップショット', () => {
    it('デフォルトの見た目が変わっていない', () => {
      const { container } = render(<ToolCard {...defaultProps} />);
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});


