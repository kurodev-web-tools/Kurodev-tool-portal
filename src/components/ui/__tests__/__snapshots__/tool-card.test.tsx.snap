// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ToolCard スナップショット デフォルトの見た目が変わっていない 1`] = `
<div
  aria-describedby="tool-1-description"
  aria-label="テストツール - これはテスト用のツールです"
  class="text-card-foreground gap-6 rounded-xl border py-6 h-full flex flex-col group card-interactive hover:shadow-blue-500/20 border-gray-800 shadow-sm bg-gray-900/30 hover:bg-gray-900/40 cursor-pointer"
  data-slot="card"
  role="button"
  tabindex="0"
>
  <div
    class="@container/card-header auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 pb-4 flex-shrink-0 hidden md:block"
    data-slot="card-header"
  >
    <div
      class="flex items-start justify-between mb-3"
    >
      <div
        class="w-12 h-12 rounded-lg bg-gradient-to-r from-blue-500 to-cyan-500 flex items-center justify-center shadow-lg text-2xl hover-scale-sm"
      >
        📅
      </div>
      <span
        class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden [a&]:hover:bg-primary/90 text-xs font-medium bg-yellow-200 text-yellow-900 border-yellow-300 dark:bg-yellow-800 dark:text-yellow-100 dark:border-yellow-600 status-beta"
        data-slot="badge"
      >
        ベータ版
      </span>
    </div>
    <div
      class="text-lg md:text-xl font-bold group-hover:text-blue-600 transition-smooth leading-tight tracking-wide"
      data-slot="card-title"
    >
      テストツール
    </div>
    <div
      class="text-xs md:text-sm text-gray-600 dark:text-gray-400 leading-relaxed"
      data-slot="card-description"
    >
      これはテスト用のツールです
    </div>
  </div>
  <div
    class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 px-6 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 pb-2 flex-shrink-0 md:hidden"
    data-slot="card-header"
  >
    <div
      class="flex items-center justify-between mb-2"
    >
      <div
        class="flex items-center space-x-3"
      >
        <div
          class="w-10 h-10 rounded-lg bg-gradient-to-r from-blue-500 to-cyan-500 flex items-center justify-center shadow-lg text-xl hover-scale-sm"
        >
          📅
        </div>
        <div>
          <div
            class="text-sm md:text-base font-bold group-hover:text-blue-600 transition-smooth leading-tight tracking-wide"
            data-slot="card-title"
          >
            テストツール
          </div>
        </div>
      </div>
      <span
        class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden [a&]:hover:bg-primary/90 text-xs font-medium bg-yellow-200 text-yellow-900 border-yellow-300 dark:bg-yellow-800 dark:text-yellow-100 dark:border-yellow-600 status-beta"
        data-slot="badge"
      >
        ベータ版
      </span>
    </div>
    <div
      class="text-xs md:text-sm text-gray-500 dark:text-gray-300 leading-relaxed"
      data-slot="card-description"
      id="tool-1-description"
    >
      これはテスト用のツールです
    </div>
  </div>
  <div
    class="px-6 flex-grow pb-4 hidden md:block"
    data-slot="card-content"
  >
    <div
      class="flex items-center gap-2 text-xs md:text-sm text-gray-400 dark:text-gray-300"
    >
      <svg
        aria-hidden="true"
        class="lucide lucide-users h-4 w-4"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M16 21v-2a4 4 0 0 0-4-4H6a4 4 0 0 0-4 4v2"
        />
        <path
          d="M16 3.128a4 4 0 0 1 0 7.744"
        />
        <path
          d="M22 21v-2a4 4 0 0 0-3-3.87"
        />
        <circle
          cx="9"
          cy="7"
          r="4"
        />
      </svg>
      <span>
        利用可能
      </span>
    </div>
  </div>
  <div
    class="px-6 flex-grow pb-2 md:hidden"
    data-slot="card-content"
  />
  <div
    class="flex items-center px-6 [.border-t]:pt-6 pt-0 flex-shrink-0"
    data-slot="card-footer"
  >
    <a
      href="/tools/test-tool"
    >
      ツールを使用する
      <svg
        aria-hidden="true"
        class="lucide lucide-arrow-right h-4 w-4 group-hover:translate-x-1 transition-smooth"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="m12 5 7 7-7 7"
        />
      </svg>
    </a>
    <a
      href="/tools/test-tool"
    >
      使用する
      <svg
        aria-hidden="true"
        class="lucide lucide-arrow-right h-3 w-3 group-hover:translate-x-1 transition-smooth"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M5 12h14"
        />
        <path
          d="m12 5 7 7-7 7"
        />
      </svg>
    </a>
  </div>
</div>
`;
